crates/sui-rosetta/src/operations.rs

```diff
@@ -0,0 +1,667 @@
+            related_operations: vec![],
+            type_: OperationType::GasBudget,
+            status,
+            account: Some(AccountIdentifier { address: sender }),
+            amount: None,
+            coin_change: Some(CoinChange {
+                coin_identifier: CoinIdentifier {
+                    identifier: gas.into(),
+                },
+                coin_action: CoinAction::CoinSpent,
+            }),
+            metadata: Some(json!({ "budget": budget })),
+        }
+    }
+}
+
+fn parse_operations(
+    tx: &SingleTransactionKind,
+    budget: u64,
+    gas: ObjectRef,
+    sender: SuiAddress,
+    counter: &mut IndexCounter,
+    status: Option<OperationStatus>,
+    effects: Option<&TransactionEffects>,
+) -> Result<Vec<Operation>, anyhow::Error> {
+    let mut operations = match tx {
+        SingleTransactionKind::TransferSui(tx) => transfer_sui_operations(
+            budget,
+            gas,
+            sender,
+            tx.recipient,
+            tx.amount,
+            counter,
+            status,
+        ),
+        SingleTransactionKind::TransferObject(tx) => transfer_object_operations(
+            budget,
+            tx.object_ref,
+            gas,
+            sender,
+            tx.recipient,
+            counter,
+            status,
+        ),
+        SingleTransactionKind::Call(c) => {
+            move_call_operations(sender, gas, budget, c, counter, status)
+        }
+        SingleTransactionKind::Publish(p) => {
+            let disassembled = disassemble_modules(p.modules.iter())?;
+            vec![Operation {
+                operation_identifier: counter.next_idx().into(),
+                related_operations: vec![],
+                type_: OperationType::Publish,
+                status,
+                account: Some(AccountIdentifier { address: sender }),
+                amount: None,
+                coin_change: None,
+                metadata: Some(json!(disassembled)),
+            }]
+        }
+        SingleTransactionKind::ChangeEpoch(change) => vec![Operation {
+            operation_identifier: counter.next_idx().into(),
+            related_operations: vec![],
+            type_: OperationType::EpochChange,
+            status,
+            account: None,
+            amount: None,
+            coin_change: None,
+            metadata: Some(json!(change)),
+        }],
+        SingleTransactionKind::Pay(pay) => parse_pay(sender, gas, budget, pay, counter, status),
+    };
+    if !matches!(tx, SingleTransactionKind::TransferSui(..)) {
+        if let Some(effects) = effects {
+            let coin_change_operations = Operation::get_coin_operation_from_event(
+                tx.input_objects()?,
+                &effects.events,
+                status,
+                counter,
+            );
+            operations.extend(coin_change_operations);
+        }
+    }
+    Ok(operations)
+}
+
+fn transfer_sui_operations(
```

Comment:
Why not just rely on the events in the effects for TransferSui?

Reply:
This is because we support TransferSui transaction creation on the Rosetta API.
One of the checks Rosetta does is compare the intents (the transaction/operation input from rosetta) with the executed transaction, it is expecting the executed transaction to contain the intents, so we need to include the TransactionData as well.

Reply:
Sorry but I didn't understand. Could you point to me the line of code where transfer_sui_operations would be generating different result than get_coin_operation_from_event for TransferSui?

Reply:
Sure, this need some explanation...

there are 2 places we convert Sui's transaction to Rosetta's operations, one is in the `/construction/parse` endpoint and another one in `/block/transaction` endpoint.

[/operation/parse](https://github.com/MystenLabs/sui/blob/2bc1596c21a22dd00d85543adb2be6793a07e101/crates/sui-rosetta/src/construction.rs#L197) uses `Operation::from_data`  to create the "intent" operations and [/block/transaction](https://github.com/MystenLabs/sui/blob/2bc1596c21a22dd00d85543adb2be6793a07e101/crates/sui-rosetta/src/block.rs#L57) uses `Operation::from_data_and_effect` to create the "confirmed" operations.

the `/construction/parse` endpoint is used for checking transaction correctness during transaction construction, in our case we only support TransferSui for now, the operations created looks like this (negative amount indicate sender):

```
              {
                "operation_identifier": {
                    "index": 0
                },
                "type": "TransferSUI",
                "account": {
                    "address": "0xc4173a804406a365e69dfb297d4eaaf002546ebd"
                },
                "amount": {
                    "value": "-100000",
                    "currency": {
                        "symbol": "SUI",
                        "decimals": 9
                    }
                },
                "coin_change": {
                    "coin_identifier": {
                        "identifier": "0x0dce9190d54cde842d39537bf94efe128181b8a6:5549"
                    },
                    "coin_action": "coin_spent"
                }
            },
            {
                "operation_identifier": {
                    "index": 1
                },
                "type": "TransferSUI",
                "account": {
                    "address": "0x96bc0b37b67103651d1f98c67b34df9558ea527a"
                },
                "amount": {
                    "value": "100000",
                    "currency": {
                        "symbol": "SUI",
                        "decimals": 9
                    }
                }
            },
            {
                "operation_identifier": {
                    "index": 2
                },
                "type": "GasBudget",
                "account": {
                    "address": "0xc4173a804406a365e69dfb297d4eaaf002546ebd"
                },
                "coin_change": {
                    "coin_identifier": {
                        "identifier": "0x0dce9190d54cde842d39537bf94efe128181b8a6:5549"
                    },
                    "coin_action": "coin_spent"
                },
                "metadata": {
                    "budget": 1000
                }
            }
```
The sender, recipients and transfer amounts are specified in the intent operations, these data are being used to create `TransactionData` for signature.

After the tx is executed, the rosetta-cli compare the intent operations with the confirmed operations , the confirmed operations must contain the intent operations (the confirmed operations can have more operations then the intent).

Since the intent operations of TransferSui contains all the balance change informations(amount field) already, we don't need to use the event to create the operations, also operation created by `get_coin_operation_from_event`  will contain recipient's coin id, which will cause a mismatch.